File: 1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.py
class Solution:
    def maxSatisfied(self, customers: List[int], grumpy: List[int], minutes: int) -> int:
        maxUnSat = 0
        curUnSat = 0
        n = len(grumpy)
        for i in range(minutes):
            curUnSat += customers[i] * grumpy[i]
        maxUnSat = curUnSat
        i = 0
        j = minutes
        while j < n:
            curUnSat += customers[j] * grumpy[j]
            curUnSat -= customers[i] * grumpy[i]
            maxUnSat = max(maxUnSat, curUnSat)
            i += 1
            j += 1
        totalSat = maxUnSat
        for k in range(n):
            totalSat += customers[k] * (1 - grumpy[k])
        return totalSat



# class Solution:
#     def maxSatisfied(self, customers: List[int], grumpy: List[int], minutes: int) -> int:
#         n = len(grumpy)
#         ans = 0
#         for i in range(n):
#             if grumpy[i] == 0:
#                 ans += customers[i]
#         temp1 = ans
#         for i in range(n-minutes+1):
#             temp = temp1
#             for j in range(i, i+minutes):
#                 if grumpy[j] == 1:
#                     temp += customers[j]
#             ans = max(ans, temp)
#         return ans

File: 1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.cpp
class Solution {
public:
    int maxSatisfied(vector<int>& customers, vector<int>& grumpy, int minutes) {
        int maxUnSat = 0;
        int curUnSat = 0;
        int n = grumpy.size();
        for (int i = 0; i < minutes; i++) {
            curUnSat += customers[i] * grumpy[i];
        }
        maxUnSat = curUnSat;
        int i = 0, j = minutes;
        while (j < n) {
            curUnSat += customers[j] * grumpy[j];
            curUnSat -= customers[i] * grumpy[i];
            maxUnSat = max(maxUnSat, curUnSat);
            i++;
            j++;
        }
        int totalSat = maxUnSat;
        for (int i = 0; i < n; i++) {
            totalSat += customers[i] * (1 - grumpy[i]);
        }
        return totalSat;
    }
};




// class Solution {
// public:
//     int maxSatisfied(vector<int>& customers, vector<int>& grumpy, int minutes) {
//         int n = grumpy.size();
//         int ans = 0;
//         for (int i = 0; i < n; i++) {
//             if (grumpy[i] == 0) {
//                 ans += customers[i];
//             }
//         }
//         int temp1 = ans;
//         for (int i = 0; i <= n - minutes; i++) {
//             int temp = temp1;
//             for (int j = i; j < i + minutes; j++) {
//                 if (grumpy[j] == 1) {
//                     temp += customers[j];
//                 }
//             }
//             ans = max(ans, temp);
//         }
//         return ans;
//     }
// };